### PR TITLE
Fix Element borders when exporting to SVG

### PR DESCRIPTION
**Pull request by AlexanderG2207**
### Checklist
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes
- [x] I translated all the newly inserted strings into German and English

### Motivation and Context
When creating a model and exporting it to an SVG, the borders of class elements are not rendered correctly (Before Image). This leads to class elements looking different when exported compared to what the user had modeled in the editor. 

#292 had an approach that reduced the element height if it was a class member. I found that the issue was regarding the border rect and path being rendered before the class members were rendered, making the class members overlap the rect and path and thus making the borders invisible. That is why a class diagram doesn't have any rendering issues with its border in the editor because the members are rendered before the rect and path (Apollon Editor Image).

### Description
To fix this issue, the render order was changed for class members. Before, each Component for the SVG was rendered by creating each element and adding it to a map (Current SVG Export Image). With UMLClassifierComponents, the UMLMemberComponents were generated after the UMLClassifierComponent, making them block the rect and paths of the UMLClassifierComponent since they are rendered in the UMLClassifierComponent.

The UMLClassifierComponent allows ReactNodes (children) to be placed within the Component. This then firstly renders the children and then secondly the rect and paths (Updated SVG Export Image). So, by first checking if the given Element is of type UMLClassifierComponent and then creating an array of all children and placing them within the UMLClassifierComponent, they are rendered before the borders. This results in the borders being visible and the SVG being exported correctly (After Image).

### Steps for Testing
1. Run Apollon locally.
2. Open Apollon in Google Chrome.
3. Create a test diagram with as many different diagram types as possible.
4. Export to SVG.
5. Take note of any border issues once the SVG is rendered.

### Test Coverage
The "class-diagram-as-svg.json" file was adjusted in the unit test resources to get all the tests to pass. This is because the whole order of SVG element rendering was changed, and therefore, the test diagram needed to be adjusted accordingly.

| File | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
|------|-------:|-----:|------:|-------:|-----:|
| svg.tsx | 81.36 | 70 | 66.66 | 81.36 | ...36-141,175-177 

### Screenshots
Apollon Editor | Current SVG Export | Updated SVG Export
:-------------------------:|:-------------------------:|:-------------------------:
Apollon Editor | Current SVG Export | Updated SVG Export

Before | After
:-------------------------:|:-------------------------:
Before | After

### CODE REVIEW

**Review by tobias-lippert — APPROVED** [review]
Tested locally with your fork. The issue seems to be fixed.

File: src/main/scenes/svg.tsx [hunk 1]
```diff
@@ -212,19 +214,56 @@ export class Svg extends Component<Props, State> {
...
+                    width={element.bounds.width}
+                    height={element.bounds.height}
+                    key={element.id}
+                    className={element.name ? element.name.replace(/[<>]/g, '') : ''}
+                    fill={element.fillColor || theme.color.background}
+                  >
+                    <ElementComponent key={index} element={element} scale={this.props.options?.scale || 1.0} />
+                  </svg>
+                );
+              }
```
Comment (pal03377) [inline src/main/scenes/svg.tsx:266]: What happens when neither of the `if` conditions are true?

File: src/main/scenes/svg.tsx [hunk 2]
```diff
@@ -212,19 +214,56 @@ export class Svg extends Component<Props, State> {
...
+                          >
+                            <MemberElementComponent key={memberIndex} element={memberElement} scale={this.props.options?.scale || 1.0} />
+                          </svg>
+                        );
+                      })}
+                    </ElementComponent>
+                  </svg>
+                );
+                // Render all other UMLElements and UMLRelationships normally, as they dont have issues when rendering to SVG
+              } else if (ElementComponent !== UMLClassifierComponent && ElementComponent !== UMLClassifierMemberComponent) {
```
Comment (pal03377) [inline src/main/scenes/svg.tsx:252]: I would move the comment into the `else`
```suggestion
              } else if (ElementComponent !== UMLClassifierComponent && ElementComponent !== UMLClassifierMemberComponent) {
                  // Render all other UMLElements and UMLRelationships normally, as they dont have issues when rendering to SVG
```

**Review by pal03377 — COMMENTED** [review]
Well done! I tested locally and didn't encounter any border issues. I only have some minor comments about the code:

File: src/main/scenes/svg.tsx [hunk 3]
```diff
@@ -212,19 +214,56 @@ export class Svg extends Component<Props, State> {
...
-                  width={element.bounds.width}
-                  height={element.bounds.height}
-                  key={element.id}
-                  className={element.name ? element.name.replace(/[<>]/g, '') : ''}
-                  fill={element.fillColor || theme.color.background}
-                >
-                  <ElementComponent key={index} element={element} scale={this.props.options?.scale || 1.0} />
-                </svg>
-              );
+              // If the ElementComponent is of type UMLClassifierComponent, create an array of all members (attributes and methods)
```
[the remark this answers is not in the document]
Comment (AlexanderG2207) [reply src/main/scenes/svg.tsx]: I added more detailed comments for each switch case :)

File: src/main/scenes/svg.tsx [hunk 4]
```diff
@@ -212,19 +214,61 @@ export class Svg extends Component<Props, State> {
...
+                  // Unlike other components, the UMLClassifierComponent needs its members to be children within the component to avoid border rendering issues.
+                  const members = elements.filter((member) => member.owner === element.id);
+                  return (
+                    <svg
+                      x={element.bounds.x}
+                      y={element.bounds.y}
+                      width={element.bounds.width}
+                      height={element.bounds.height}
+                      key={element.id}
+                      className={element.name ? element.name.replace(/[<>]/g, '') : ''}
```
Comment (florian-glombik) [inline src/main/scenes/svg.tsx:229]: This happens 3 times within this component, might be worth it to use a method `convertElementNameToClass` or something similar to improve readability and reduce code duplication?

**Review by florian-glombik — APPROVED** [review]
Tested it locally, and the issue is resolved for me.

The code looks good!

I did have the same issue within `apollon-editor.ts` as described in https://github.com/ls1intum/Apollon/pull/298#pullrequestreview-1605732048, not sure why this only seems to appear for me :sweat_smile:

File: src/main/scenes/svg.tsx (hunk 4 above)
Comment (AlexanderG2207) [reply src/main/scenes/svg.tsx]: Good idea! I created a method for the duplicate code.

**Pull request merged**